include/votca/xtp/bseoperator_btda.h

```diff
@@ -0,0 +1,232 @@
+  typedef double Scalar;
+  typedef double RealScalar;
+  typedef int StorageIndex;
+  enum {
+    ColsAtCompileTime = Eigen::Dynamic,
+    MaxColsAtCompileTime = Eigen::Dynamic,
+    IsRowMajor = false
+  };
+
+  HamiltonianOperator(const MatrixReplacementA &A,
+                      const MatrixReplacementB &B) 
+  : _A(A), _B(B) 
+  { 
+    _size = 2*A.cols() ;
+    _diag = get_diagonal();
+   };
+      
+    
+  class InnerIterator {
+   public:
+    InnerIterator(const HamiltonianOperator& xpr, const Eigen::Index& id)
+        : _xpr(xpr), _id(id){};
+
+    InnerIterator& operator++() {
+      _row++;
+      return *this;
+    }
+    operator bool() const {
+      return _row < _xpr._size;
+    }  
+    
+    double value() const { return _xpr(_row, _id); }
+    Eigen::Index row() const { return _row; }
+    Eigen::Index col() const { return _id; }
+    Eigen::Index index() const { return row(); }
+
+   private:
+    const HamiltonianOperator& _xpr;
+    const Eigen::Index _id;
+    Eigen::Index _row = 0;
+  };
+
+  Eigen::Index rows() const { return this->_size; }
+  Eigen::Index cols() const { return this->_size; }
+  Eigen::Index outerSize() const { return this->_size; }
+
+  template <typename Vtype>
+  Eigen::Product<HamiltonianOperator, Vtype, Eigen::AliasFreeProduct>
+      operator*(const Eigen::MatrixBase<Vtype>& x) const {
+    return Eigen::Product<HamiltonianOperator, Vtype,
+                          Eigen::AliasFreeProduct>(*this, x.derived());
+  }
+
+  // this is not a fast method
+  const double& operator()(const size_t i, const size_t j) const {
+    if (i==j) {
+      return _diag(i);
+    } else {
+      Eigen::RowVectorXd row_out = row(i);
+      return  row_out(j);
+    }
+  };
+
+  //  get a row of the operator
+  Eigen::RowVectorXd row(int index) const {
+    /* Returna row of the operator 
+    H = [ A   B
+         -B* -A* ]
+    */
+    int lsize = this->_size;
+    int halfsize = lsize/2;
+    Eigen::RowVectorXd row_out = Eigen::RowVectorXd::Zero(lsize);
+
+    if (index < halfsize)
+    {
+
+      Eigen::RowVectorXd a = this->_A.row(index);
+      Eigen::RowVectorXd b = this->_B.row(index);    
+      row_out.head(halfsize) = a;
+      row_out.tail(halfsize) = b;
+
+    } else {
+
+      Eigen::RowVectorXd a = this->_A.row(index-halfsize);
+      Eigen::RowVectorXd b = this->_B.row(index-halfsize);    
+      row_out.head(halfsize) = -b;
+      row_out.tail(halfsize) = -a;
+    }
+
+    return row_out;
+  }
+
+  Eigen::VectorXd get_diagonal() const {
+    Eigen::VectorXd diag = Eigen::VectorXd::Zero(_size);
+#pragma omp parallel for
+    for (int i=0; i < _size; i++) {
+      Eigen::RowVectorXd r = row(i);
+      diag(i) = r(i);
+    }
+    return diag;
+  }
+
+  Eigen::VectorXd diagonal() const {return _diag;}
+  
+  // get the full matrix if we have to
+  Eigen::MatrixXd get_full_matrix() const {
+    Eigen::MatrixXd matrix = Eigen::MatrixXd::Zero(_size, _size);
+  
+    for (int i = 0; i < _size; i++) {
```

Comment:
matrix.topleftcorner=A.fullmatrix()
matrix.toprightcorner=B.fullmatrix()
matrix.bottomleftcorner=-matrix.toprightcorner.transpose()
matrix.bottomrightcorner=-matrix.topleftcorner()

Reply:
no need for rows

Reply:
done